### PR TITLE
Add Duration Prediction Model

### PR DESCRIPTION
**Pull request by JarodMica**
Worked to implement duration model as trained and implemented here by lucasnewman here: https://github.com/lucasnewman/f5-tts-mlx/tree/main

Tested and it works with infer-cli.py, would love to have others test and verify as I'm no PyTorch wizard.

### CODE REVIEW

File: src/f5_tts/infer/examples/basic/basic.toml [hunk 1]
```diff
@@ -7,4 +7,5 @@ gen_text = "I don't really care what you call me. I've been a silent spectator,
 # File with text to generate. Ignores the text above.
 gen_file = ""
 remove_silence = false
-output_dir = "tests"
\ No newline at end of file
+output_dir = "tests"
+predict_duration = false
```
Comment (SWivid) [inline src/f5_tts/infer/examples/basic/basic.toml:11]: `predict_duration` -> duration_model` for clearer definition of var
same for following codes

File: src/f5_tts/infer/infer_cli.py [hunk 2]
```diff
@@ -16,6 +16,7 @@
     load_vocoder,
     preprocess_ref_audio_text,
     remove_silence_for_generated_wav,
+    load_prediction_model
```
Comment (SWivid) [inline src/f5_tts/infer/infer_cli.py:19]: `load_prediction_model` -> `load_duration_model`
same for following codes

File: src/f5_tts/infer/infer_cli.py [hunk 3]
```diff
@@ -81,6 +82,13 @@
     default=1.0,
     help="Adjust the speed of the audio generation (default: 1.0)",
 )
+
+parser.add_argument(
+    "--predict_duration",
+    type=bool,
+    default=False,
+    help="Automatically predict the duration of spoken segment"
```
Comment (SWivid) [inline src/f5_tts/infer/infer_cli.py:90]: help="Leverage a duration model to predict the duration"

File: src/f5_tts/infer/utils_infer.py [hunk 4]
```diff
@@ -5,33 +5,35 @@
...
+)
+from f5_tts.model import CFM
+from vocos import Vocos
+from transformers import pipeline
+from pydub import AudioSegment, silence
+import tqdm
+import torchaudio
+import torch
+import numpy as np
+import matplotlib.pylab as plt
```
Comment (SWivid) [inline src/f5_tts/infer/utils_infer.py:20]: `from xxx impor xxx` go after `import xxx`
and better in alphabet order of pkg name first letter
same for following codes

File: src/f5_tts/infer/utils_infer.py [hunk 5]
```diff
@@ -5,33 +5,35 @@
...
-from f5_tts.model import CFM
-from f5_tts.model.utils import (
-    get_tokenizer,
-    convert_char_to_pinyin,
-)
 
 _ref_audio_cache = {}
 
-device = "cuda" if torch.cuda.is_available() else "mps" if torch.backends.mps.is_available() else "cpu"
+device = "cuda" if torch.cuda.is_available(
```
Comment (SWivid) [inline src/f5_tts/infer/utils_infer.py:35]: will need pre-commit formatting,
refer to https://github.com/SWivid/F5-TTS?tab=readme-ov-file#development
same for following code

p.s. just need run the pre-commit once, before `git commit`

File: src/f5_tts/infer/utils_infer.py [hunk 6]
```diff
@@ -359,7 +466,7 @@ def infer_batch_process(
     nfe_step=32,
     cfg_strength=2.0,
     sway_sampling_coef=-1,
-    speed=1,
+    speed=0.8,
```
Comment (SWivid) [inline src/f5_tts/infer/utils_infer.py:469]: better go for the linear estimated duration operation part (a factor there) if serves as default

File: src/f5_tts/infer/utils_infer.py [hunk 7]
```diff
@@ -388,12 +495,23 @@ def infer_batch_process(
         ref_audio_len = audio.shape[-1] // hop_length
         if fix_duration is not None:
             duration = int(fix_duration * target_sample_rate / hop_length)
+        elif prediction_model:
+            duration_in_sec = prediction_model(audio, text_list)
+            print(duration_in_sec)
+            frame_rate = model_obj.mel_spec.sample_rate // model_obj.mel_spec.hop_length
```
Comment (SWivid) [inline src/f5_tts/infer/utils_infer.py:501]: `model_obj.mel_spec.sample_rate` -> `model_obj.mel_spec.target_sample_rate`
duplicate definition of var

File: src/f5_tts/model/duration.py [hunk 8]
```diff
@@ -0,0 +1,358 @@
+# adapted from https://github.com/lucasnewman/f5-tts-mlx/blob/main/f5_tts_mlx/duration.py
```
Comment (SWivid) [inline src/f5_tts/model/duration.py:1]: duration class should go into modules.py
and leverage existing modules there (cuz it's all the same for attn, ffn, melspec, pos_emb, etc.)

File: src/f5_tts/model/mlx_modules.py [hunk 9]
```diff
@@ -0,0 +1,512 @@
+from __future__ import annotations
```
Comment (SWivid) [inline src/f5_tts/model/mlx_modules.py:1]: probably no separate mlx_modules.py in this repo,
leveraging existing modules.py is better for repo maintainability

File: src/f5_tts/model/modules.py [hunk 10]
```diff
@@ -118,6 +118,7 @@ def __init__(
         self.hop_length = hop_length
         self.win_length = win_length
         self.n_mel_channels = n_mel_channels
+        self.sample_rate = target_sample_rate
```
Comment (SWivid) [inline src/f5_tts/model/modules.py:121]: remove: cuz duplicate def for target_sample_rate

File: src/f5_tts/model/utils.py [hunk 11]
```diff
@@ -69,8 +69,10 @@ def mask_from_frac_lengths(seq_len: int["b"], frac_lengths: float["b"]):  # noqa
 def maybe_masked_mean(t: float["b n d"], mask: bool["b n"] = None) -> float["b d"]:  # noqa: F722
     if not exists(mask):
         return t.mean(dim=1)
-
-    t = torch.where(mask[:, :, None], t, torch.tensor(0.0, device=t.device))
+    
+    mask = mask.to(device=t.device, dtype=torch.bool)
+    mask = mask.unsqueeze(-1) 
+    t = torch.where(mask, t, torch.zeros_like(t))
```
Comment (SWivid) [inline src/f5_tts/model/utils.py:75]: no sure, if critical to modify this part

File: src/f5_tts/train/finetune_gradio.py [hunk 12]
```diff
@@ -1372,7 +1372,7 @@ def get_audio_select(file_sample):
 with gr.Blocks() as app:
```
Comment (SWivid) [inline src/f5_tts/train/finetune_gradio.py:1372]: Very nice updates here for clearer wording and phrase standardization !!

**Review by SWivid — CHANGES_REQUESTED** [review]
Thanks for PR !!

**Comment by lpscr** [comment]
@JarodMica great thank you very much
i just try to test i ger error

```
git clone https://github.com/JarodMica/F5-TTS.git
git submodule update --init --recursive # this i am not sure if need but i install also
pip install e .
```

run
`python src/f5_tts/infer/infer_cli.py`

```
Traceback (most recent call last):
  File "c:\PythonApps\f5ttsbg\F5-TTS\src\f5_tts\infer\infer_cli.py", line 234, in <module>
    main()
  File "c:\PythonApps\f5ttsbg\F5-TTS\src\f5_tts\infer\infer_cli.py", line 230, in main
    main_process(ref_audio, ref_text, gen_text, ema_model, mel_spec_type, remove_silence, speed,  prediction_model)
NameError: name 'prediction_model' is not defined. Did you mean: 'load_prediction_model'?
```

what i do wrong ?

**Comment by lpscr** [comment]
@SWivid  @JarodMica 
results
https://vocaroo.com/1i1i9zSO1x46

**Comment by lpscr** [comment]
here i make test clear repo for begin install 

value :
seed = 0
gen_text = "i would like to test the new value for predicting duration. I think it could improve our results, and I'm interested to see how it performs."

results

@SWivid main repo
https://vocaroo.com/1md4yDJNqNeX

@JarodMica with predict_duration 
https://vocaroo.com/19CAUVRTX2Tv

**Comment by lpscr** [comment]
@JarodMica today i just create new clear install 

i use in base model and finetune and working great ! 


`pip install torch==2.4.0 torchvision==0.19.0 torchaudio==2.4.0 --index-url https://download.pytorch.org/whl/cu118`

```
git clone https://github.com/JarodMica/F5-TTS.git
git submodule update --init --recursive # I'm not sure if I need this, but I also installed
pip install e .
```

fix in `utils_infer.py`

```
def infer_process(
    ref_audio,
    ref_text,
    gen_text,
    model_obj,
    vocoder, # here
    prediction_model=None, #here
   ...
):
```

run
`python src/f5_tts/infer/infer_cli.py  --predict_duration True`

if you have error with miss lib 
`packages\torch\lib\fbgemm.dll" or one of its dependencies. `
download and copy the libomp140.x86_64.dll file in  /windows/system32

i make some change in `api.py` and you can use 


```
f5tts = F5TTS(predict_duration=False)

wav, sr, spect = f5tts.infer(
    ref_file="basic_ref_en.wav",
    ref_text="some call me nature, others call me mother nature.",
    gen_text="""I don't really care what you call me. I've been a silent spectator, watching species evolve, empires rise and fall. But always remember, I am mighty and enduring. Respect me and I'll nurture you; ignore me and you shall face the consequences."""
)

```

```
import random
import sys
from importlib.resources import files

import soundfile as sf
import torch
import tqdm
from cached_path import cached_path

from f5_tts.infer.utils_infer import (
    hop_length,
    infer_process,
    load_model,
    load_vocoder,
    preprocess_ref_audio_text,
    remove_silence_for_generated_wav,
    save_spectrogram,
    load_prediction_model,
    target_sample_rate,
)
from f5_tts.model import DiT, UNetT
from f5_tts.model.utils import seed_everything


class F5TTS:
    def __init__(
        self,
        model_type="F5-TTS",
        ckpt_file="",
        vocab_file="",
        ode_method="euler",
        use_ema=True,
        vocoder_name="vocos",
        local_path=None,
        predict_duration=True,
        device=None,
    ):
        # Initialize parameters
        self.final_wave = None
        self.target_sample_rate = target_sample_rate
        self.hop_length = hop_length
        self.seed = -1
        self.mel_spec_type = vocoder_name

        # Set device
        self.device = device or (
            "cuda" if torch.cuda.is_available() else "mps" if torch.backends.mps.is_available() else "cpu"
        )
        

        if predict_duration:
           self.prediction_model=load_prediction_model()
        else:   
           self.prediction_model=None

        # Load models
        self.load_vocoder_model(vocoder_name, local_path)
        self.load_ema_model(model_type, ckpt_file, vocoder_name, vocab_file, ode_method, use_ema)

    def load_vocoder_model(self, vocoder_name, local_path):
        self.vocoder = load_vocoder(vocoder_name, local_path is not None, local_path, self.device)

    def load_ema_model(self, model_type, ckpt_file, mel_spec_type, vocab_file, ode_method, use_ema):
        if model_type == "F5-TTS":
            if not ckpt_file:
                if mel_spec_type == "vocos":
                    ckpt_file = str(cached_path("hf://SWivid/F5-TTS/F5TTS_Base/model_1200000.safetensors"))
                elif mel_spec_type == "bigvgan":
                    ckpt_file = str(cached_path("hf://SWivid/F5-TTS/F5TTS_Base_bigvgan/model_1250000.pt"))
            model_cfg = dict(dim=1024, depth=22, heads=16, ff_mult=2, text_dim=512, conv_layers=4)
            model_cls = DiT
        elif model_type == "E2-TTS":
            if not ckpt_file:
                ckpt_file = str(cached_path("hf://SWivid/E2-TTS/E2TTS_Base/model_1200000.safetensors"))
            model_cfg = dict(dim=1024, depth=24, heads=16, ff_mult=4)
            model_cls = UNetT
        else:
            raise ValueError(f"Unknown model type: {model_type}")

        self.ema_model = load_model(
            model_cls, model_cfg, ckpt_file, mel_spec_type, vocab_file, ode_method, use_ema, self.device
        )

    def export_wav(self, wav, file_wave, remove_silence=False):
        sf.write(file_wave, wav, self.target_sample_rate)

        if remove_silence:
            remove_silence_for_generated_wav(file_wave)

    def export_spectrogram(self, spect, file_spect):
        save_spectrogram(spect, file_spect)

    def infer(
        self,
        ref_file,
        ref_text,
        gen_text,
        show_info=print,
        progress=tqdm,
        target_rms=0.1,
        cross_fade_duration=0.15,
        sway_sampling_coef=-1,
        cfg_strength=2,
        nfe_step=32,
        speed=1.0,
        fix_duration=None,
        remove_silence=False,
        file_wave=None,
        file_spect=None,
        seed=-1,
    ):
        if seed == -1:
            seed = random.randint(0, sys.maxsize)
        seed_everything(seed)
        self.seed = seed

        ref_file, ref_text = preprocess_ref_audio_text(ref_file, ref_text, device=self.device)

        wav, sr, spect = infer_process(
            ref_file,
            ref_text,
            gen_text,
            self.ema_model,
            self.vocoder,
            self.prediction_model,
            self.mel_spec_type,
            show_info=show_info,
            progress=progress,
            target_rms=target_rms,
            cross_fade_duration=cross_fade_duration,
            nfe_step=nfe_step,
            cfg_strength=cfg_strength,
            sway_sampling_coef=sway_sampling_coef,
            speed=speed,
            fix_duration=fix_duration,
            device=self.device,
        )

        if file_wave is not None:
            self.export_wav(wav, file_wave, remove_silence)

        if file_spect is not None:
            self.export_spectrogram(spect, file_spect)

        return wav, sr, spect
```

File: src/f5_tts/infer/utils_infer.py (hunk 5 above)
Comment (JarodMica) [reply src/f5_tts/infer/utils_infer.py]: I was having some formatting issues I guess with it:

```
src\f5_tts\model\modules.py:746:18: F722 Syntax error in forward annotation: `b n d`
    |
744 |     def __call__(
745 |         self,
746 |         x: float["b n d"],
    |                  ^^^^^^^ F722
747 |         text_embed: float["b n d"],
748 |     ):
    |
    
   ```
   
   which were present already in the repo before.  I committed without verification, but it reformatted the repo, got any suggestions?

File: src/f5_tts/model/utils.py (hunk 11 above)
Comment (JarodMica) [reply src/f5_tts/model/utils.py]: Changed back, proved non-critical

**Comment by JarodMica** [comment]
@lpscr Thanks for testing!  Pushed the latest repo with the updates and suggestions from @SWivid, so please test that if you have some time.  I've also incorporated your suggestions into the API but updated with the names that are being used

**Comment by JarodMica** [comment]
Also, with torch at 2.4.0, there were issues with whisperx/cttranslate being incompatible with it a few months back.  That area may need some testing to see if it works correctly on 2.4.0 @lpscr .  I'll get try it out sometime later to see if that has been resolved

**Comment by lpscr** [comment]
@JarodMica Yes! I'm going to test your new update I just need to finish something first. i tell you soon

**Comment by lpscr** [comment]
@JarodMica ok i goind to test now

**Comment by JarodMica** [comment]
> @JarodMica Yes! I'm going to test your new update I just need to finish something first. i tell you soon

It's all good, no rush. I'm going to bed, it's 5 in the morning lol.

**Comment by lpscr** [comment]
@JarodMica i just install with 

`pip install torch==2.3.0+cu118 torchaudio==2.3.0+cu118 --extra-index-url https://download.pytorch.org/whl/cu118`
error again :
`AttributeError: module 'torch.nn' has no attribute 'RMSNorm'`

so install 
`pip install torch==2.3.0+cu118 torchaudio==2.3.0+cu118 --extra-index-url https://download.pytorch.org/whl/cu118`

and working again fine api and cli !

from src.f5_tts.api import F5TTS
from IPython.display import Audio, display

f5tts = F5TTS(duration_model=True)

api
```
from src.f5_tts.api import F5TTS
from IPython.display import Audio, display

f5tts = F5TTS(duration_model=True)

wav, sr, spect = f5tts.infer(
    ref_file="basic_ref_en.wav",
    ref_text="some call me nature, others call me mother nature.",
    gen_text="I don't really care what you call me. I've been a silent spectator, watching species evolve, empires rise and fall. But always remember, I am mighty and enduring." 
)

```
cli
`python src/f5_tts/infer/infer_cli.py  --duration_model True  `

**Comment by lpscr** [comment]
> > @JarodMica Yes! I'm going to test your new update I just need to finish something first. i tell you soon
> 
> It's all good, no rush. I'm going to bed, it's 5 in the morning lol.

No problem at all! Get some rest; sleep well!

File: src/f5_tts/infer/utils_infer.py (hunk 5 above)
Comment (SWivid) [reply src/f5_tts/infer/utils_infer.py]: similar as in current modules.py file, you could do 
`# noqa: F722`

Comment (JarodMica) [reply src/f5_tts/infer/utils_infer.py]: Thanks, never use pre-commit before.  Looks like something I'll get used to using, good tool.

**Review by JarodMica — COMMENTED** [review]
Appropriate changes made

**Comment by SWivid** [comment]
Hi @JarodMica , Thanks a lot for this PR~
There are actually some problems that we need to discuss:

1. We are probably releasing new models with different tokenizer/vocab, which will forcing training new duration predictors
2. How is the latency of duration model which is a crucial trade-off part cuz our goal is a speech-to-speech realtime application
3. As we all know the duration model follow the flow-matching NAR paradigm, which is not streaming, and probably will be abandoned for reasons similar to 2.
4. Consider point 2., we actually tried with providing groundtruth duration info for model to do evaluation. As you can see from our paper, the WER (indicates how faithful the TTS output is with groundtruth) already surpass or close to groundtruth one, and the speech slider (e.g. 1.0 -> 0.8) is quite straightforward to do adjustment while the duration predictor results tend to ignore pauses style in audio prompt

So generally we are kind of consider take the duration predictor off. 
We should just speak frankly and sincerely earlier at the start, cuz it's taking you much time and effort to incorporate the duration prediction model here.
We always and will always be grateful to everyone (and especially you guys) in the community for working hard for open source stuff. And we may have the version with duration predictor in a forked repo as an stable version, which is also fantastic.

Best~

**Comment by JarodMica** [comment]
> 2. How is the latency of duration model which is a crucial trade-off part cuz our (final) goal is a speech-to-speech realtime application

It adds on about .25-.30 seconds to total generation time which is the same across different lengths of outputted audio.

> 1. We are probably releasing new models with different tokenizer/vocab, which will forcing training new duration predictors

I'm looking into how long it would take to finetune the duration predictor for new langauges/new vocab, but it doesn't look to be too bad.  

I think the biggest issue is sometimes linear prediction comes off as unnatural if you leave at default, therefore, I do have to adjust the speech slider manually which would get a bit tedious for bulk generation.  In many cases, I prefer the output of prediction being used vs linear est, but that's my preference so a user may be able to choose between using it or not using it based on what they like. Would it be more fitting then I guess to call it a speed predictor?  Because the main benefit I see is needing to rely on changing the speed slider less and more on training the model to predict it.

basic_example:
No dur pred: https://voca.ro/1k3e60QEoFf0
dur pred used: https://voca.ro/165MG3NYptVG

Out of dis speaker:
No dur pred: https://voca.ro/1lh4mluAbCPQ
dur pred: https://voca.ro/13llRW2rsw6a

I would be more than happy with keeping it on my fork and just maintaining it there as I do have plans on incorporating this repo in other projects!

**Comment by SWivid** [comment]
> Would it be more fitting then I guess to call it a speed predictor?

Yes, as it serves more like a sentence-level duration predictor compared to previous phoneme-level duration predictor which is providing more prior assumptions while limiting naturalness upper bound, like that.

> basic_example:

It's actually also out of dis ref_audio lol, and we are actually working on better solution for linear estimated dur compared to current way.

**Comment by lpscr** [comment]
@JarodMica hi in Out of dis speaker: in begin you have glitsh effect noise no good in ear 

i have found solution days now for this but i trouble with other stuff i want make pull request 

to fix simple just you you need clear any puase in audio end and begin then add just 50ms pause in end for me working great !
test to see if this working  i test in finetune model only but must working in base model if working let me know 

if this working it's easy fix and with have very nice clear audio

here code i use in `utils_infer.py`

```
def remove_silence_from_both_ends(audio, silence_threshold=-42):
    # Remove silence from the start
    non_silent_start_duration = 0
    for segment in audio:
        if segment.dBFS > silence_threshold:
            break
        non_silent_start_duration += segment.duration_seconds
    audio = audio[int(non_silent_start_duration * 1000):]
    
    # Remove silence from the end
    non_silent_end_duration = audio.duration_seconds
    for segment in reversed(audio):
        if segment.dBFS > silence_threshold:
            break
        non_silent_end_duration -= segment.duration_seconds
    trimmed_audio = audio[:int(non_silent_end_duration * 1000)]
    
    return trimmed_audio
```

https://github.com/SWivid/F5-TTS/blob/61ff2a62d9487e3362ffa5680007e788ad764065/src/f5_tts/infer/utils_infer.py#L260
add in line 261 and 262

```
aseg = remove_silence_from_both_ends(aseg)
aseg = aseg + AudioSegment.silent(duration=50) 
```

before
image
after
image

**Comment by JarodMica** [comment]
> @JarodMica hi in Out of dis speaker: in begin you have glitsh effect noise no good in ear
> 
> i have found solution days now for this but i trouble with other stuff i want make pull request
> 
> to fix simple just you you need clear any puase in audio end and begin then add just 50ms pause in end for me working great ! test to see if this working i test in finetune model only but must working in base model if working let me know

It definitely works, I just got a bit lazy with my audio here: https://github.com/SWivid/F5-TTS/issues/85#issuecomment-2438903529

Appreciate the advise though!

**Comment by lpscr** [comment]
@JarodMica great i make pull request thank you for the test

**Pull request closed without merging**